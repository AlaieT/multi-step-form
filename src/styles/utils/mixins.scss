@mixin flex($param: null, $type: null) {
  display: flex;

  @if  $param == "center&center" {
    justify-content: center;
    align-items: center;
  }

  @if  $param == "center&end" {
    justify-content: center;
    align-items: flex-end;
  }

  @if  $param == "end&center" {
    justify-content: flex-end;
    align-items: center;
  }

  @if  $param == "space-between&center" {
    justify-content: space-between;
    align-items: center;
  }

  @if  $type == "col" {
    flex-direction: column;
  }
}
