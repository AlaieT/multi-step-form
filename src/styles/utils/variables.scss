$background-white: rgb(255 255 255);
$background-global: rgb(235 243 254);
$background-select-step: rgb(191 226 253);
$primary-text-color: rgb(2 41 90);
$secondary-text-color: rgb(150 153 171);
$plan-border-color: rgb(85 75 209);
$summary-hover-button: rgb(133 127 250);
$select-input: rgb(71 61 255);
$select-button: rgb(0 62 138);
$select-plan: rgb(246 247 254);
$error: hsl(354deg 84% 57%);
$shadow-color: rgb(214 217 230);
$content-box-shadow: 0 25px 40px -20px rgba(0 0 0 / 10%);
