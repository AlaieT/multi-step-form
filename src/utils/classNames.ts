const classNames = (...classes: (string | undefined)[]): string | undefined => {
  return classes.filter((item) => !!item).join(" ");
};

export default classNames;
