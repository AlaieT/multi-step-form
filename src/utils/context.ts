import { createContext } from "react";

import type { FormContextType } from "../types";

const FormContext = createContext<FormContextType>({
  form: {
    planMode: "yearly",
    priceMode: "yr"
  },
  setForm: () => undefined
});

export { FormContext };
