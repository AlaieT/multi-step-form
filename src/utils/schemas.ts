import { object, string, boolean, number } from "yup";

const multiStepFromSchema = object({
  name: string().required("Name is a required field!"),
  email: string()
    .required("Email is a required field!")
    .email("Must be a valid email!"),
  phone: string().required("Phone is a required field!"),
  plan: string().oneOf(["arcade", "advanced", "pro"]).defined(),
  planMode: boolean().defined(),
  onlineService: boolean().defined(),
  largeStorage: boolean().defined(),
  customizableProfile: boolean().defined(),
  totalPrice: number().defined(),
});

export { multiStepFromSchema };
