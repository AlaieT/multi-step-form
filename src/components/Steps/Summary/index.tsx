import React, { useContext, useEffect } from "react";

import Button from "../../Button";
import Wrap from "../Wrap";

import type { SummaryProps } from "../../../types";

import styles from "../../../styles/components/steps/summary.module.scss";
import { FormContext } from "../../../utils/context";

const Summary = ({
  plans,
  addOns,
  isValid,
  setValue,
  getValues,
  changeStep
}: SummaryProps) => {
  const { form } = useContext(FormContext);
  const planType = getValues("plan");

  const totalPrice = plans[planType][form.planMode]
    + Object.keys(addOns)
      .map((key) => (getValues(key) ? addOns[key][form.planMode] : 0))
      .reduce((prev, curr) => prev + curr, 0);

  useEffect(() => {
    setValue("totalPrice", totalPrice, { shouldValidate: false });
    return () => undefined;
  }, [setValue, totalPrice]);

  return (
    <Wrap>
      <div id={styles.content}>
        <div id={styles.header}>
          <h1>Finishing up</h1>
          <h2>Double-check everything looks OK before confirming.</h2>
        </div>
        <div id={styles.info}>
          <div id={styles.plan}>
            <p>
              {planType}
              {" "}
              {form.planMode}
            </p>
            <Button id={styles.change_plan} onClick={() => changeStep(1)}>
              Change
            </Button>
            <p>{`$${plans[planType][form.planMode]}/${form.priceMode}`}</p>
          </div>
          {getValues("onlineService") && (
            <div className={styles.add_ons}>
              <span>Online Service</span>
              <span>
                {`+ ${addOns.onlineService[form.planMode]}/${form.priceMode}`}
              </span>
            </div>
          )}
          {getValues("largeStorage") && (
            <div className={styles.add_ons}>
              <span>Large Storage</span>
              <span>
                {`+ ${addOns.largeStorage[form.planMode]}/${form.priceMode}`}
              </span>
            </div>
          )}
          {getValues("customizableProfile") && (
            <div className={styles.add_ons}>
              <span>Customizable Profile</span>
              <span>
                {`+ ${addOns.customizableProfile[form.planMode]}/${form.priceMode}`}
              </span>
            </div>
          )}
        </div>
        <div id={styles.total}>
          <span>Total (per year)</span>
          <span>{`$${totalPrice}/${form.priceMode}`}</span>
        </div>
      </div>
      <div id={styles.controls}>
        <Button id={styles.go_back} type="button" onClick={() => changeStep(2)}>
          Go Back
        </Button>
        <Button id={styles.next_step} type="submit" disabled={!isValid}>
          Confirm
        </Button>
      </div>
    </Wrap>
  );
};

export default Summary;
