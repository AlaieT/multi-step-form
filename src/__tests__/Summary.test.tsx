import React from "react";
import { render, cleanup, renderHook } from "@testing-library/react";

import Summary from "../components/Steps/Summary";
import { useForm } from "react-hook-form";
import { FormContext } from "../utils/context";

afterEach(cleanup);

describe("Summary", () => {
  const props = global.getStepProps();
  const mockChangeStep = jest.fn();

  describe("correct render", () => {
    it("should match snapshot", () => {
      const { result } = renderHook(() => useForm(global.getUseFormProps()));

      expect(
        render(
          <FormContext.Provider
            value={{
              form: { planMode: "yearly", priceMode: "yr" },
              setForm: () => undefined
            }}
          >
            <Summary
              plans={props.plans}
              addOns={props.addOns}
              isValid
              setValue={result.current.setValue}
              getValues={result.current.getValues}
              changeStep={mockChangeStep}
            />
          </FormContext.Provider>
        ).asFragment()
      ).toMatchSnapshot();
    });
  });
});
