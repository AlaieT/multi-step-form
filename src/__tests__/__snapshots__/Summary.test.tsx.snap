// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary correct render should match snapshot 1`] = `
<DocumentFragment>
  <div
    id="step"
  >
    <div
      id="content"
    >
      <div
        id="header"
      >
        <h1>
          Finishing up
        </h1>
        <h2>
          Double-check everything looks OK before confirming.
        </h2>
      </div>
      <div
        id="info"
      >
        <div
          id="plan"
        >
          <p>
            arcade yearly
          </p>
          <button
            class="button"
            id="change_plan"
            type="button"
          >
            Change
          </button>
          <p>
            $90/yr
          </p>
        </div>
      </div>
      <div
        id="total"
      >
        <span>
          Total (per year)
        </span>
        <span>
          $90/yr
        </span>
      </div>
    </div>
    <div
      id="controls"
    >
      <button
        class="button"
        id="go_back"
        type="button"
      >
        Go Back
      </button>
      <button
        class="button"
        id="next_step"
        type="submit"
      >
        Confirm
      </button>
    </div>
  </div>
</DocumentFragment>
`;
